pkg/middlewares/ratelimiter/redisrate/rate.go

```diff
@@ -0,0 +1,151 @@
+package redisrate
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"strconv"
+	"time"
+
+	"github.com/redis/go-redis/v9"
+	"golang.org/x/time/rate"
+)
+
+const redisPrefix = "rate:"
+
+// Inf is the infinite rate limit; it allows all events (even if burst is zero).
+const Inf = rate.Limit(math.MaxFloat64)
+
+// InfDuration is the duration returned by Delay when a Reservation is not OK.
+const InfDuration = time.Duration(math.MaxInt64)
+
+type rediser interface {
+	Eval(ctx context.Context, script string, keys []string, args ...interface{}) *redis.Cmd
+	EvalSha(ctx context.Context, sha1 string, keys []string, args ...interface{}) *redis.Cmd
+	ScriptExists(ctx context.Context, hashes ...string) *redis.BoolSliceCmd
+	ScriptLoad(ctx context.Context, script string) *redis.StringCmd
+	Del(ctx context.Context, keys ...string) *redis.IntCmd
+
+	EvalRO(ctx context.Context, script string, keys []string, args ...interface{}) *redis.Cmd
+	EvalShaRO(ctx context.Context, sha1 string, keys []string, args ...interface{}) *redis.Cmd
+}
+
+type Limit struct {
+	Rate   rate.Limit
+	Burst  int64
+	Period time.Duration
+}
+
+func (l Limit) String() string {
+	return fmt.Sprintf("%f req/%s (burst %d)", l.Rate, fmtDur(l.Period), l.Burst)
+}
+
+func (l Limit) IsZero() bool {
+	return l == Limit{}
+}
+
+func fmtDur(d time.Duration) string {
+	switch d {
+	case time.Second:
+		return "s"
+	case time.Minute:
+		return "m"
+	case time.Hour:
+		return "h"
+	}
+	return d.String()
+}
+
+func PerSecond(rate rate.Limit) Limit {
+	return Limit{
+		Rate:   rate,
+		Period: time.Second,
+		Burst:  int64(rate),
+	}
+}
+
+func PerMinute(rate rate.Limit) Limit {
+	return Limit{
+		Rate:   rate,
+		Period: time.Minute,
+		Burst:  int64(rate),
+	}
+}
+
+func PerHour(rate rate.Limit) Limit {
+	return Limit{
+		Rate:   rate,
+		Period: time.Hour,
+		Burst:  int64(rate),
+	}
+}
+
+// ------------------------------------------------------------------------------
+
+// Limiter controls how frequently events are allowed to happen.
+type Limiter struct {
+	rdb      rediser
+	ttl      int
+	maxDelay time.Duration
+}
+
+// NewLimiter returns a new Limiter.
+func NewLimiter(rdb rediser, ttl int, maxDelay time.Duration) *Limiter {
```

Comment:
As stated in my first review, the `RedisLimiter` is thin enough that I think merging this `redisrate.Limiter and `RedisLimiter` will help maintainability. Was there a reason I missed for this split?

Reply:
I keep this in a separate package since `redisrate` will be in charge of executing the script and handles logic based on the limiting algorithm like the `x/time/rate` does

Reply:
But on second thought I am thinking about merging back to RedisLimiter for simplizing this.